src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java

```diff
@@ -543,8 +550,20 @@ public RemoteAction buildRemoteAction(Spawn spawn, SpawnExecutionContext context
         platform = PlatformUtils.getPlatformProto(spawn, remoteOptions);
       }
 
+      var useOutputPaths = true;
+      if (mayBeExecutedRemotely(spawn)) {
```

Comment:
What's the concern for not checking the server capabilities for cache-only actions?

My understanding is we should use the same message format for both remote cache and remote execution.

Reply:
Or maybe we should remove the check. The problem now is we only store the server capabilities in `remoteExecutor` which could be `null` in case of remote cache only build. Can we store the server capabilities in `RemoteExecutionService` instead?

Reply:
I think we could switch to storing ServerCapabilities inside RemoteCache instead of CacheCapabilities.
That way we could access it here using the logic I proposed above, like this

```java
  private boolean shouldUseOutputPaths(Spawn spawn) {
    var isRemoteExec = mayBeExecutedRemotely(spawn);
    var isRemoteCached = mayBeCachedRemotely(spawn);
    if (!isRemoteExec || !isRemoteCached) {
      return true;
    }

    ServerCapabilities serverCapabilities;
    if (isRemoteExec) {
      serverCapabilities = remoteExecutor.getServerCapabilities();
    } else {
      serverCapabilities = remoteCache.getServerCapabilities();
    }
    if (serverCapabilities == null) {
      return true;
    }

    var supportStatus = ClientApiVersion.current.checkServerSupportedVersions(serverCapabilities);
    if (supportStatus.isUnsupported()) {
      return true;
    }

    return supportStatus.getHighestSupportedVersion().compareTo(ApiVersion.twoPointOne) >= 0;
  }
```

This solution should work with the exception to the case where the RemoteCache and the RemoteExecutor support 2 different low/high version pairs in their capability. Using the logic above ☝️  we would bias to using Executor capabilities instead of RemoteCache capabilities. But a future change could be made so that we could use the lowest supported version among the 2.

WDYT? I could send out a separate PR for the RemoteCache refactoring to `s/CacheCapabilities/ServerCapabilities/g`, or include that as part of this PR as a separate commit.

Reply:
That could also work. Please create a separate PR for the RemoteCache refactoring. Thanks!

Reply:
I sent https://github.com/bazelbuild/bazel/pull/18734 for that.

I also rebase this PR on top of that PR and included https://github.com/bazelbuild/bazel/pull/18270/commits/c5b59e65ed4f2256c2ea7b8a9da34757f068ee23 to demonstrate how the end result would look like.

Reply:
@coeuvre Thanks for this pointer. I forgot about this thread.

I went back to #18734 and found it mostly no longer relevant because of the recent refactoring in master branch. I have added the needed API as a separate commit here though.

This makes the current PR look like this
```
> git log --oneline -6 --reverse
c08353126d RemoteCacheClient: add getServerCapabilities
afce68f4b8 Apply google-java-format to RemoteWorkerOptions
11c09c134f Conditionally set output_paths based on Remote Executor capabilities
c43cd180a8 init useoutputpaths
2de7c2d8f9 RemoteOptions: add useOutputPaths
2d7f14fb6a RemoteOptions: flip remote_use_output_paths
```

The first 5 commits could be backported to 7.3.0 if needed. The 6th commit, which flip the flag value, could be an 8.0-only change.
I have tested the first 5 commits exclusively to make sure they pass CI. Though, I have to admit, the additional flag made testing a bit sparse. I hope it's short-lived enough for that to not matter.